File: Weekly_1_C/Problem_Set_1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number;
    int sum = 0;
    int isSecond = 0;

    // Credit card number
    number = get_long("Number: ");

    long originalNumber = number;

    // Counter for digits of number
    int count = 0;
    while (originalNumber > 0)
    {
        count++;
        originalNumber /= 10;
    }

    // Reset
    originalNumber = number;

    // Luhn's Algorithm
    while (number > 0)
    {
        int digit = number % 10;

        if (isSecond)
        {
            digit = digit * 2;
            if (digit > 9)
            {
                digit = digit - 9;
            }
        }

        sum += digit;
        isSecond = !isSecond;
        number /= 10;
    }

    // Check the first two digits (or the first digit in the case of VISA) of the original number
    while (originalNumber >= 100)
    {
        originalNumber /= 10;
    }

    if (sum % 10 == 0)
    {
        if ((originalNumber >= 40 && originalNumber < 50) && (count == 13 || count == 16))
        {
            printf("VISA\n");
        }
        else if ((originalNumber >= 51 && originalNumber <= 55) && count == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((originalNumber == 34 || originalNumber == 37) && count == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Weekly_6_Python/Problem_Set_6/credit.py
from cs50 import get_int


def main():
    nmb = get_int("Number: ")
    p_nmb = nmb

    count = 0
    while p_nmb > 0:
        count += 1
        p_nmb //= 10

    p_nmb = nmb
    sum = 0
    scnd = False

    while nmb > 0:
        digit = nmb % 10
        if scnd:
            digit *= 2
            if digit > 9:
                digit -= 9

        sum += digit
        scnd = not scnd
        nmb //= 10

    while p_nmb >= 100:
        p_nmb //= 10

    if sum % 10 == 0:
        if (p_nmb // 10 == 4) and (count == 13 or count == 16):
            print("VISA")
        elif (p_nmb >= 51 and p_nmb <= 55) and count == 16:
            print("MASTERCARD")
        elif (p_nmb in [34, 37]) and count == 15:
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
